### PR TITLE
[DO NOT MERGE] add support for multilib mpi

### PR DESCRIPTION
**Pull request by junghans**

### CODE REVIEW

File: sys-cluster/mpich/mpich-3.2-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# Copyright 1999-2016 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Id$
+
+EAPI=5
```
Comment (SoapGentoo) [inline sys-cluster/mpich/mpich-3.2-r1.ebuild]: Any reason not to go for `EAPI=6`?

File: sys-cluster/mpich/mpich-3.2-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	/usr/include/mpi.h
+	/usr/include/opa_config.h
+)
+
+pkg_setup() {
+	FORTRAN_STANDARD="77 90"
+	fortran-2_pkg_setup
+
+	if use mpi-threads && ! use threads; then
+		ewarn "mpi-threads requires threads, assuming that's what you want"
```
Comment (SoapGentoo) [inline sys-cluster/mpich/mpich-3.2-r1.ebuild]: Imo, this is too much implicit inference. We should rather go for `REQUIRED_USE="mpi-threads? ( threads )"`, and then we can chuck this whole `pkg_setup()` function.

File: sys-cluster/mpich/mpich-3.2-r1.ebuild [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	fi
+}
+
+src_prepare() {
+	# Using MPICHLIB_LDFLAGS doesn't seem to fully work.
```
Comment (SoapGentoo) [inline sys-cluster/mpich/mpich-3.2-r1.ebuild]: will need `default`

File: sys-cluster/mpich/mpich-3.2-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		c="${c} --with-thread-package=pthreads"
+		c="${c} --enable-threads=runtime"
+	else
+		if use threads ; then
+			c="${c} --with-thread-package=pthreads"
+		else
+			c="${c} --with-thread-package=none"
+		fi
+		c="${c} --enable-threads=single"
+	fi
```
Comment (SoapGentoo) [inline sys-cluster/mpich/mpich-3.2-r1.ebuild]: while whole section can be made redundant with two `usex` lines once `REQUIRED_USE` is used

File: sys-cluster/mpich/mpich-3.2-r1.ebuild [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+	c="${c} --sysconfdir=${EPREFIX}/etc/${PN}"
+	c="${c} --docdir=${EPREFIX}/usr/share/doc/${PF}"
```
Comment (SoapGentoo) [inline sys-cluster/mpich/mpich-3.2-r1.ebuild]: don't attach these to `${c}`, rather attach them to `econf` directly, clearer, less indirection

File: sys-cluster/mpich/mpich-3.2-r1.ebuild [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+	dodir /usr/share/doc/${PF}
+	dodoc README{,.envvar} CHANGES RELEASE_NOTES
+	newdoc src/pm/hydra/README README.hydra
+	if use romio; then
+		newdoc src/mpi/romio/README README.romio
+	fi
+
+	if ! use doc; then
+		rm -rf "${D}"usr/share/doc/${PF}/www*
```
Comment (SoapGentoo) [inline sys-cluster/mpich/mpich-3.2-r1.ebuild]: missing `|| die`

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	echo
+}
+
+src_prepare() {
+	# Necessary for scalibility, see
+	# http://www.open-mpi.org/community/lists/users/2008/09/6514.php
+	if use threads; then
+		echo 'oob_tcp_listen_mode = listen_thread' \
+			>> opal/etc/openmpi-mca-params.conf
+	fi
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: missing `|| die`

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 8]
```diff
@@ -0,0 +1,201 @@
+# Copyright 1999-2016 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Id$
+
+EAPI=5
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: there are more eclasses involved here, still, did you try `EAPI=6`?

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+		--enable-orterun-prefix-by-default
+		--with-hwloc="${EPREFIX}/usr"
+		--with-libltdl="${EPREFIX}/usr"
+		)
+
+	if use fortran; then
+		myconf+=(--enable-mpi-fortran=all)
+	else
+		myconf+=(--enable-mpi-fortran=no)
+	fi
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: replace block with `usex` instead

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+	if use java; then
+		# We must always build with the right -source and -target
+		# flags. Passing flags to javac isn't explicitly supported here
+		# but we can cheat by overriding the configure test for javac.
+		export ac_cv_path_JAVAC="$(java-pkg_get-javac) $(java-pkg_javac-args)"
+	fi
+
+	! use vt && myconf+=(--enable-contrib-no-build=vt)
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: same here, prefer `usex` instead

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+		$(multilib_native_use_enable openmpi_ofed_features_failover btl-openib-failover) \
+		$(multilib_native_use_with openmpi_rm_pbs tm) \
+		$(multilib_native_use_with openmpi_rm_slurm slurm)
+}
+
+multilib_src_install() {
+	default
+
+	# Remove la files, no static libs are installed and we have pkg-config
+	find "${ED}"/usr/$(get_libdir)/ -type f -name '*.la' -delete
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: prefer `find "${ED}" -name '*.la' -delete || die`

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+		mv "${ED}"/usr/include/mpif* "${T}"/fortran || die
+	else
+		#some fortran files get installed unconditionally 
+		rm "${ED}"/usr/include/mpif* "${ED}"/usr/bin/mpif* || die
+	fi
+}
+
+multilib_src_install_all() {
+	# From USE=vt see #359917
+	rm "${ED}"/usr/share/libtool &> /dev/null
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: missing `|| die`

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+	# fortran header cannot be wrapped (bug #540508), workaround part 2
+	if use fortran; then
+		mv "${T}"/fortran/mpif* "${ED}"/usr/include || die
+	fi
+
+	# Avoid collisions with libevent
+	rm -rf "${ED}"/usr/include/event2 &> /dev/null
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: missing `|| die`

File: virtual/mpi/mpi-2.0-r4.ebuild [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# Copyright 1999-2016 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Id$
+
+EAPI=5
```
Comment (SoapGentoo) [inline virtual/mpi/mpi-2.0-r4.ebuild]: `EAPI=6`?

File: sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+multilib_src_configure() {
+	local myconf=(
+		--sysconfdir="${EPREFIX}/etc/${PN}"
+		--enable-pretty-print-stacktrace
+		--enable-orterun-prefix-by-default
+		--with-hwloc="${EPREFIX}/usr"
+		--with-libltdl="${EPREFIX}/usr"
+		)
```
Comment (SoapGentoo) [inline sys-cluster/openmpi/openmpi-1.10.2-r1.ebuild]: attach all of this stuff directly to `econf`, seeing the next block is not needed anymore

**Comment by junghans** [comment]
Sorry, for now please keep your review to the multilib parts.

**Comment by SoapGentoo** [comment]
@junghans The changes look good otherwise. If you want I can also merge the PR and add the necessary edits.

**Comment by junghans** [comment]
If you like the multilib part, take it and do the rest.

**Comment by SoapGentoo** [comment]
@junghans I performed the fixups. Maybe give the whole lot a go?

**Pull request closed without merging**